src/coreComponents/linearAlgebra/interfaces/hypre/HypreVector.cpp

```diff
@@ -459,16 +320,16 @@ void HypreVector::print( std::ostream & os ) const
     MpiWrapper::barrier( getComm() );
     if( rank == myRank )
     {
-      real64 const * const local_data = extractLocalVector();
+      arrayView1d< real64 const > const data = values();
       globalIndex const firstRowID = ilower();
-      for( localIndex i = 0; i < localSize(); ++i )
+      forAll< serialPolicy >( localSize(), [&, data]( localIndex const i )
```

Comment:
I think this would be more useful if it printed out in the desired memory space. We would of course have to use `printf` for this....but I often want to see the values on device rather than the host values.

The thing is that you have to launch a bunch of single threaded kernels in sequence to get the output in order when on device....unless there is a better way to do this.

Reply:
The intention is that view semantics always does the right thing and gets you the most recently modified values. What would be the use case for printing directly from device? Device values will only be different if we forget to call `touch()` in one of the LAI methods... I think? But in that case seeing outdated values printed might even help in discovering the bug IMO.

There's also the fact that current API prints to an iostream, which we can't do with `printf`... which if course we could change and make it always go to `stdout`. Still, not a fan of either `printf` (`%d` vs `%lld` thing especially) or launching a cuda kernel per element.

Reply:
I view it as a diagnostic tool to make sure you have what you think you have in the place you think it should be. Since the underlying object is managed by `LvArray`, but hypre will not respect the touching/moving semantics. It will just do what it wants. So being able to print both the host and device copies has been useful for me at times.